impl/maven-core/src/main/java/org/apache/maven/internal/transformation/impl/PomInlinerTransformer.java

```diff
@@ -52,35 +61,79 @@ class PomInlinerTransformer extends TransformerSupport {
+                                    return (String)
+                                            session.getConfigProperties().get(property);
+                                }
+                                return null;
+                            },
+                            false);
+                    if (!Objects.equals(originalPom, interpolatedPom)) {
+                        Files.writeString(tmpPom, interpolatedPom);
+                        artifact = artifact.setPath(tmpPom);
+                    }
+                } catch (IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+            }
+            newArtifacts.add(artifact);
+        }
+        return newArtifacts;
+    }
+
+    @SuppressWarnings("unchecked")
+    private Set<String> needsInlining(RepositorySystemSession session) {
+        return (Set<String>) session.getData()
+                .computeIfAbsent(
+                        PomInlinerTransformer.class.getName() + ".needsInlining", ConcurrentHashMap::newKeySet);
+    }
+
     @Override
     public void injectTransformedArtifacts(RepositorySystemSession session, MavenProject project) throws IOException {
         if (!Features.consumerPom(session.getConfigProperties())) {
             try {
                 Model model = read(project.getFile().toPath());
-                boolean parentVersion = false;
                 String version = model.getVersion();
                 if (version == null && model.getParent() != null) {
-                    parentVersion = true;
                     version = model.getParent().getVersion();
                 }
                 String newVersion;
                 if (version != null) {
+                    HashSet<String> usedProperties = new HashSet<>();
                     newVersion = interpolator.interpolate(version.trim(), property -> {
                         if (!session.getConfigProperties().containsKey(property)) {
                             throw new IllegalArgumentException("Cannot inline property " + property);
                         }
+                        usedProperties.add(property);
                         return (String) session.getConfigProperties().get(property);
                     });
                     if (!Objects.equals(version, newVersion)) {
-                        if (parentVersion) {
-                            model = model.withParent(model.getParent().withVersion(newVersion));
-                        } else {
-                            model = model.withVersion(newVersion);
-                        }
-                        Path tmpPom = Files.createTempFile(
-                                project.getArtifactId() + "-" + project.getVersion() + "-", ".xml");
-                        write(model, tmpPom);
-                        project.setFile(tmpPom.toFile());
+                        needsInlining(session).addAll(usedProperties);
```

Comment:
yes.

Reply:
might give dedication as well, if makes sense.
```suggestion
                        needsInlining(session).addAll(getUsedProperties(interpolator));
```
might be not performant as n+1 iteration.

because the code does something like simple `map` operation on kind collection interpolator.interpolate(). Idk this structure so well im sorry. Somehow can not reach the code, so my knowledge very short.